### PR TITLE
ROX-27203: Ensure collector/scanner images exist during updates

### PR DESCRIPTION
**Pull request by tommartensen**
### Description

Ensure that GHA and Konflux built images are present when attempting to update the SCANNER_VERSION, COLLECTOR_VERSION files.
Had to bump COLLECTOR_VERSION to a successful Konflux build. 

### User-facing documentation

- [x] CHANGELOG  update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request)  is not needed

### Testing and quality

- ~[ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag~
- [ ] CI results are inspected

#### Automated testing

- ~[ ] added unit tests~
- ~[ ] added e2e tests~
- ~[ ] added regression tests~
- ~[ ] added compatibility tests~
- ~[ ] modified existing tests~

#### How I validated my change

* check-collector-images-exist
* check-scanner-images-exist 

checks on the build pipeline.

### CODE REVIEW

File: .github/workflows/update_collector_periodic.yaml [hunk 1]
```diff
@@ -4,21 +4,46 @@ on:
...
+    strategy:
+      matrix:
+        tag_suffix: ["", "-fast"]
+    steps:
+    - name: Ensure image exists
+      uses: stackrox/actions/release/wait-for-image@v1
+      with:
+        token: ${{ secrets.QUAY_RHACS_ENG_BEARER_TOKEN }}
+        image: rhacs-eng/collector:${{ needs.determine-new-collector-version.outputs.new-collector-version }}${{ matrix.tag_suffix }}
+        limit: 1800
```
Comment (msugakov) [inline .github/workflows/update_collector_periodic.yaml:34]: What happens when this fails? Does the scheduled workflow get quietly failing without notifying anyone via Slack and without updating the existing PRs or opening the new ones?

If yes, it would be hard to notice when PRs stop updating in the volume of all other auto-generated PRs.

Also, this check here doesn't catch when someone manually updates SCANNER_VERSION/COLLECTOR_VERSION in their PR.

I'd suggest to introduce the check for images existence in one of the PR/test workflows and have it conditionally enabled only when SCANNER_VERSION or COLLECTOR_VERSION get changed.

This way we both catch the case when PRs opened by this automation try bump pointers to commits for which Konflux builds failed and also prevent manual updates to broken builds, and the check status will be visible to engineers.

WDYT?

Comment (tommartensen) [reply .github/workflows/update_collector_periodic.yaml]: Moved the check to `build.yaml` as `check-collector-images-exist ` and `check-scanner-images-exist` - as a prerequisite for the image build. 
This will be reported as part of the build pipeline (which is a required check) either as a failed check on a PR or as a Slack message on pushes with the existing mechanisms.

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -251,9 +251,54 @@ jobs:
...
+    - name: Checkout repo
+      uses: actions/checkout@v4
+
+    - name: Get COLLECTOR_VERSION
+      id: collector-version
+      run: |
+        echo "collector-version=$(make --quiet --no-print-directory collector-tag)" >> "${GITHUB_OUTPUT}"
+
+    - name: Check image exists
+      uses: stackrox/actions/release/wait-for-image@v1
```
[the remark this answers is not in the document]
Comment (tommartensen) [reply .github/workflows/build.yaml:269]: Yes, that is correct: https://github.com/stackrox/actions/releases/tag/v1
We auto-release on that repository and move the `v1` tag to the new location.

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -251,9 +251,54 @@ jobs:
...
+    steps:
+    - name: Checkout repo
+      uses: actions/checkout@v4
+
+    - name: Get COLLECTOR_VERSION
+      id: collector-version
+      run: |
+        echo "collector-version=$(make --quiet --no-print-directory collector-tag)" >> "${GITHUB_OUTPUT}"
+
+    - name: Check image exists
```
[the remark this answers is not in the document]
Comment (tommartensen) [reply .github/workflows/build.yaml:268]: It would be possible to add steps that use [tj-actions/changed-files](https://github.com/tj-actions/changed-files), BUT it would require more work to handle both PRs and branch push differently.
Unlikely, though not impossible: the scanner or collector images could be cleaned up by Quay's image retention, which would fail this job and require the engineer to bump the versions.
The job is a prerequisite for image builds and runs in parallel to longer-running compilation jobs, so doesn't add to the overall build time.
There is no monetary cost accrued on GHA for ACS to run this job. 

TLDR: there is no harm to running the check on each build, so why overengineer it.

File: .github/workflows/build.yaml [hunk 4]
```diff
@@ -251,9 +251,54 @@ jobs:
...
+      uses: stackrox/actions/release/wait-for-image@v1
+      with:
+        token: ${{ secrets.QUAY_RHACS_ENG_BEARER_TOKEN }}
+        image: rhacs-eng/collector:${{ steps.collector-version.outputs.collector-version }}${{ matrix.tag_suffix }}
+        limit: 1800
```
[the remark this answers is not in the document]
Comment (tommartensen) [reply .github/workflows/build.yaml]: Done in https://github.com/stackrox/stackrox/pull/13502/commits/aad7eace6bdb0616ff2d9a0a14e9d8fb462ed2bb

File: .github/workflows/build.yaml [hunk 5]
```diff
@@ -251,9 +251,54 @@ jobs:
           path: |
             image/rhel/docs
 
+  check-collector-images-exist:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        tag_suffix: ["", "-fast", "-fast-slim"]
```
[the remark this answers is not in the document]
Comment (tommartensen) [reply .github/workflows/build.yaml:260]: Done in https://github.com/stackrox/stackrox/pull/13502/commits/aad7eace6bdb0616ff2d9a0a14e9d8fb462ed2bb

File: .github/workflows/build.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (msugakov) [reply .github/workflows/build.yaml:271]: Ok, I see https://github.com/stackrox/actions/blob/v1/.github/workflows/move-tag.yml

Moving git tags is surely possible but I'm not sure it is considered "canonical". I would suggest the same can be solved by introducing something like `release` branch and doing `git reset --hard <new-ref>` on it in the similar workflow. It is much more common that branches are updated than tags.

File: .github/workflows/build.yaml [hunk 6]
```diff
@@ -252,6 +252,8 @@ jobs:
             image/rhel/docs
 
   check-collector-images-exist:
+    # This job ensures that COLLECTOR_VERSION cannot be updated to a version for which
+    # the image was not successfully built on GHA and Konflux.
```
Comment (msugakov) [inline .github/workflows/build.yaml:256]: If we're to really check GHA, we should include `quay.io/stackrox-io/` repos. Like [I mentioned earlier](https://github.com/stackrox/stackrox/pull/13502#discussion_r1876090857), we did not need this at all when GHA was the only upstream build CI. I suggest removing the claim about GHA (here and for scanner job).

```suggestion
    # the image was not successfully built on Konflux.
```

File: .github/workflows/build.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (msugakov) [reply .github/workflows/build.yaml:271]: Obviously, this is something for your consideration and not blocking this PR.

File: .github/workflows/build.yaml [hunk 7]
```diff
@@ -251,9 +251,58 @@ jobs:
...
+    - name: Check image exists
+      uses: stackrox/actions/release/wait-for-image@v1
+      with:
+        token: ${{ secrets.QUAY_RHACS_ENG_BEARER_TOKEN }}
+        image: rhacs-eng/collector:${{ steps.collector-version.outputs.collector-version }}${{ matrix.tag_suffix }}
+        limit: 300
+
+  check-scanner-images-exist:
+    # This job ensures that SCANNER_VERSION cannot be updated to a version for which
+    # the image was not successfully built on GHA and Konflux.
```
Comment (msugakov) [inline .github/workflows/build.yaml:279]: There's one to remove here.

```suggestion
    # the image was not successfully built on Konflux.
```

**Review by msugakov — APPROVED** [review]
